federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+
+		// TODO: We could do a dry-run here if wanted to avoid the spurious cloud-calls & events when we restart
+
+		// The DNS records don't exist yet, so create them.
+		// TODO: quinton: Split this up and handle each record separately.
+		//                Beware though, creating distant ones before near ones can lead to race condidtions on name resolution.
+		//                (e.g. client local resolution fails as DNS record doesn't exist yet, but then tries remote resolution and that succeeds,
+		//                so it gets stuck for a while on a remote service shard.
+		//                Solutions: either use transactions (where providers suport them, e.g. Google Cloud DNS) or just make
+		//                sure that the creation sequence is zone-local, region-local, sub-continent local etc.  The latter probably
+		//                works for all cases, so transactions are probably not necessary.
+		//                Actually, on further thought, the race condition exists anyway - neither of the above tow really solve it.
+		//                Ho hum, have to think harder.
+		s.eventRecorder.Event(service, api.EventTypeNormal, "CreatingDNSRecords", "Creating DNS Records")
+		err := s.ensureDNSRecords(clusterName, service)
+		if err != nil {
+			return fmt.Errorf("Failed to create DNS records for service %s: %v", key, err), retryable
+		}
+		s.eventRecorder.Event(service, api.EventTypeNormal, "CreatedDNSRecords", "Created DNS records")
+	}
+
+	serviceInMap, ok := s.cache.get(service.Name)
+	if !ok {
+		return fmt.Errorf("service cache lost")
+	}
+	// TODO: remove dup ip info?
+	serviceInMap.appliedState.Spec.ExternalIPs = append(serviceInMap.appliedState.Spec.ExternalIPs, service.Spec.ExternalIPs)
+	// write back to federation apiserver
+
+	// TODO to check LB info equals
+	//if !api.LoadBalancerStatusEqual(previousState, &service.Status.LoadBalancer) { // TODO: quinton: only interested in the DNS record here.
+	//	// Not even sure it needs updating - check.
+	//} else {
+	_, err := s.federationClient.Core().Services().Update(serviceInMap)
+	if err != nil {
+		return fmt.Errorf("fail to update federation service")
+	}
+	//	glog.V(2).Infof("Not persisting unchanged LoadBalancerStatus to registry.") // TODO: quinton: check/fix this.
+	//}
+
+	return nil, notRetryable
+}
+
+// Returns whatever error occurred along with a boolean indicator of whether it
+// should be retried.
+func (s *ServiceController) ensureDNSRecordsExist(key string, fedService, appliedState *v1.Service) (error, bool) {
+	// Clone federation service, and create them in underlying k8s cluster
+	clone, err := conversion.NewCloner().DeepCopy(fedService)
+	if err != nil {
+		return nil, err
+	}
+	service, ok := clone.(*v1.Service)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected service cast error : %v\n", service)
+	}
+	// to generate service name, we need a api.ObjectMeta instead of v1
+	meta := &api.ObjectMeta{}
+	meta.GenerateName = service.ObjectMeta.Name + "-"
```

Comment:
@quinton-hoole need to draw a conclusion here:
@nikhiljindal and I had a call regarding to how to create kubernetes service without name conflict and there were two options, the code here is tending to implement option 2, generated random name before creating kubernetes service and use label "service-spec-hash" to link that.
If we do not generate a unique, then we do not need the label anymore and go for option 1, right? 

```
The first approach we discussed was to create k8s services with the same name as the federation service. But then what if a k8s service already exists with that name? Do we say federated service failed? How do we represent that?

The second approach we discussed was to use labels. Similar to deployments, we use a label whose value is hash of ServiceSpec. We add this label to the federation service. ServiceController in its sync loop looks at all federationServices, then for each federation service, asks each k8s cluster for services with that same label. If it gets back a service, all is well and it does nothing, else it creates a service in that k8s cluster with that label.
```